### PR TITLE
Laboratory work #4, Ekaterina Khomutova - 22FPL2

### PR DESCRIPTION
**Pull request by EkaterinaKhomutova**

### CODE REVIEW

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +28,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
+        words = text.lower().split()
+        tokens = []
+        for word in words:
+            end_of_sentence ='!?.'
```
[the remark this answers is not in the document]
Comment (EkaterinaKhomutova) [reply lab_4_fill_words_by_ngrams/main.py:36]: fixed

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,6 +28,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+            if word.isalpha() and word[-1] not in end_of_sentence:
```
[the remark this answers is not in the document]
Comment (EkaterinaKhomutova) [reply lab_4_fill_words_by_ngrams/main.py:37]: Why? In the first condition there are just words, but in the second one there could be only last words with a punctuation sign in the end. And also we need to check if our string consists of letters (not numbers or signs in string format)

[the remark this answers is not in the document]
Comment (marina-kaz) [reply lab_4_fill_words_by_ngrams/main.py]: if the first condition is true then the second one is definitely true as well
if it is false, the second may be true or false, yeah

i guess i meant they overlap a lot

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -98,6 +134,34 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+        if not (
+                isinstance(seq_len, int) and seq_len > 0 and
+                isinstance(prompt, str) and prompt
+        ):
+            raise ValueError
+        encoded = self._word_processor.encode(prompt)
+        if not encoded:
+            raise ValueError
+        #
+        for i in range(seq_len):
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:146]: if the variable is not used anywhere, name it `_`

**Review by marina-kaz — COMMENTED** [review]
Looks fine!

File: lab_4_fill_words_by_ngrams/main.py (hunk 3 above)
Comment (EkaterinaKhomutova) [reply lab_4_fill_words_by_ngrams/main.py:146]: fixed

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request merged**